server/src/metrics/edge_metrics.rs

```diff
@@ -0,0 +1,329 @@
+use ahash::{HashMap};
+use chrono::{DateTime, Utc};
+use serde::{Deserialize, Serialize};
+use tracing::info;
+use ulid::Ulid;
+use utoipa::ToSchema;
+
+use crate::types::BuildInfo;
+
+#[derive(Debug, Default, Clone, Copy, Deserialize, Serialize, ToSchema)]
+#[serde(rename_all = "camelCase")]
+pub struct LatencyMetrics {
+    pub avg: f64,
+    pub count: f64,
+    pub p99: f64,
+}
+
+impl LatencyMetrics {
+    pub fn new() -> Self {
+        Self {
+            avg: 0.0,
+            count: 0.0,
+            p99: 0.0,
+        }
+    }
+}
+
+pub const DESIRED_URLS: [&str; 6] = ["/api/client/features", "/api/client/metrics", "/api/client/metrics/bulk","/api/client/metrics/edge","/api/frontend", "/api/proxy"];
+
+#[derive(Debug, Default, Clone, Copy, Deserialize, Serialize, ToSchema)]
+#[serde(rename_all = "camelCase")]
+pub struct ProcessMetrics {
+    pub cpu_usage: f64,
+    pub memory_usage: f64,
+}
+
+#[derive(Debug, Default, Clone, Deserialize, Serialize, ToSchema)]
+#[serde(rename_all = "camelCase")]
+pub struct InstanceTraffic {
+    pub get: HashMap<String, LatencyMetrics>,
+    pub post: HashMap<String, LatencyMetrics>,
+    pub access_denied: HashMap<String, LatencyMetrics>,
+}
+
+#[derive(Debug, Default, Clone, Deserialize, Serialize, ToSchema)]
+#[serde(rename_all = "camelCase")]
+pub struct UpstreamLatency {
+    pub features: LatencyMetrics,
+    pub metrics: LatencyMetrics,
+    pub edge: LatencyMetrics,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, ToSchema)]
+#[serde(rename_all = "camelCase")]
+pub struct EdgeInstanceData {
+    pub identifier: String,
+    pub app_name: String,
+    pub region: Option<String>,
+    pub edge_version: String,
+    pub process_metrics: Option<ProcessMetrics>,
+    pub started: DateTime<Utc>,
+    pub traffic: InstanceTraffic,
+    pub latency_upstream: UpstreamLatency,
+    pub connected_streaming_clients: u64,
+    pub connected_edges: Vec<EdgeInstanceData>,
+}
+
+impl EdgeInstanceData {
+    pub fn new(app_name: &str) -> Self {
+        let build_info = BuildInfo::default();
+        Self {
+            identifier: Ulid::new().to_string(),
+            app_name: app_name.to_string(),
+            region: std::env::var("AWS_REGION").ok(),
+            edge_version: build_info.package_version.clone(),
+            process_metrics: None,
+            started: Utc::now(),
+            traffic: InstanceTraffic::default(),
+            latency_upstream: UpstreamLatency::default(),
+            connected_edges: Vec::new(),
```

Comment:
Not married to it but I'm so used to seeing the macro invocation over explicit new here

```suggestion
            connected_edges: vec![],
```

Reply:
yeah, fixed